core/src/subgraph/instance_manager.rs

```diff
@@ -1,17 +1,40 @@
+use futures::future::{loop_fn, Loop};
 use futures::sync::mpsc::{channel, Receiver, Sender};
-use graph::data::subgraph::schema::SubgraphDeploymentEntity;
-use graph::prelude::{SubgraphInstance as SubgraphInstanceTrait, *};
 use std::collections::HashMap;
+use std::sync::atomic::{AtomicBool, Ordering};
 use std::sync::RwLock;
 use std::time::Duration;
+use uuid::Uuid;
+
+use graph::data::subgraph::schema::{
+    DynamicEthereumContractDataSourceEntity, SubgraphDeploymentEntity,
+};
+use graph::prelude::{SubgraphInstance as SubgraphInstanceTrait, *};
+use graph::web3::types::Log;
 
 use super::SubgraphInstance;
 use crate::elastic_logger;
 use crate::split_logger;
 use crate::ElasticDrainConfig;
 use crate::ElasticLoggingConfig;
 
-type InstanceShutdownMap = Arc<RwLock<HashMap<SubgraphDeploymentId, CancelGuard>>>;
+struct SubgraphState<B, S, T>
+where
+    B: BlockStreamBuilder,
+    S: Store + ChainStore,
+    T: RuntimeHostBuilder,
+{
+    pub deployment_id: SubgraphDeploymentId,
+    pub instance: Arc<RwLock<SubgraphInstance<T>>>,
```

Comment:
I'm not sure how that would work. We need to read and write from the subgraph instance (read to call `matches_log` and `process_log`, write to add dynamic runtime hosts). And since there's async processing where we send the subgraph instance to another Tokio task for processing logs, I don't see an easy way around `Arc<RwLock` or `Arc<Mutex<`.

Reply:
This `Arc<RwLock<` is the one thing I still have a gripe with, I think we can remove it entirely, I'll make a new review.